docs/proposals/docker-registry-split.md

```diff
@@ -0,0 +1,141 @@
+# Move docker-registry to separate repository
+
+## Problem
+
+The `docker/distribution` has a lot of dependencies. Since `kubernetes` began to use docker/distribution
+it has become very difficult to update `docker/distribution` in the part where you need to combine the
+dependencies of these projects. There is no guarantee that the merge of the dependencies would not make
+regression in these projects.
+
+## Constraints and Assumptions
+
+The docker-registry is highly integrated in a common code-base so it is very difficult to describe
+all the difficulties of separation.
+
+## Investigation
+
+### etcd
+
+The docker-registry does not have its own `etcd`. All metadata are stored to the `origin` database using
+client API.
+
+### code-base
+
+The `origin` interacts with the docker-registry only through client library (no internal calls),
+but this cannot be said about docker-registry.
+
+1. docker-registry should use [clientsets](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/generating-clientset.md)
+of `origin`.
+2. docker-registry should use [clientsets](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/generating-clientset.md)
+of `kubernetes`.
+3. We need to create versioned client (clientsets?) for docker-registry as well.
+4. We need to split some utilities to reduce the number of dependencies to `origin`.
+5. We need to make common framework for tests. Currently docker-registry uses a common test utilities and clients.
+
+### testing
+
+Use versioning clients will allow us maintain compatibility between repositories. In this case,
+in the jenkins job for `origin` we can build the image from the separate repository before build and
+test `origin` itself. Right now we make docker-registry image from the `origin` repository.
```

Comment:
this should also reduce the size of docker-registry right? (when we start not importing stuff from origin).

do we also need a regression tests (iow. registry vs. old version of origin).